File: internal/challenge/catalog.go
package challenge

func DefaultCatalog() []Challenge {
	return []Challenge{
		{
			ID:                "calibrate-frame",
			Title:             "Frame Calibration",
			Category:          "calibration",
			Mechanic:          "hold-visible",
			Description:       "Stand in frame until landmarks are stable and visible.",
			TargetSeconds:     8,
			MinimumScore:      0.72,
			MinimumVisibility: 0.68,
			Cues: []string{
				"Keep head, shoulders, hips, knees, and ankles visible.",
				"Step back until the full body outline is inside the capture lane.",
			},
			Unlocks: []Unlock{
				{ID: "trail-overlay", Label: "Motion trail overlay", Kind: "visual"},
			},
		},
		{
			ID:                "mirror-squat",
			Title:             "Mirror Squat",
			Category:          "movement",
			Mechanic:          "squat-reps",
			Description:       "Complete controlled squats while keeping both knees tracked.",
			TargetReps:        5,
			MinimumScore:      0.76,
			MinimumVisibility: 0.64,
			Cues: []string{
				"Keep feet planted and knees visible.",
				"Move slowly enough for the tracker to follow the full range.",
			},
			Unlocks: []Unlock{
				{ID: "avatar-wireframe", Label: "Wireframe avatar", Kind: "avatar"},
			},
		},
		{
			ID:                "balance-beacon",
			Title:             "Balance Beacon",
			Category:          "control",
			Mechanic:          "balance-hold",
			Description:       "Hold a centered pose with low jitter for the target time.",
			TargetSeconds:     12,
			MinimumScore:      0.82,
			MinimumVisibility: 0.70,
			Cues: []string{
				"Center your hips over the guide line.",
				"Keep shoulders level and breathe normally.",
			},
			Unlocks: []Unlock{
				{ID: "precision-export", Label: "Precision export preview", Kind: "export"},
			},
		},
		{
			ID:                "overhead-signal",
			Title:             "Overhead Signal",
			Category:          "gesture",
			Mechanic:          "hands-up-hold",
			Description:       "Raise both hands above your head and hold the signal cleanly.",
			TargetSeconds:     5,
			MinimumScore:      0.74,
			MinimumVisibility: 0.62,
			Cues: []string{
				"Lift both wrists above your nose line.",
				"Keep shoulders visible so the gesture can be confirmed.",
			},
			Unlocks: []Unlock{
				{ID: "signal-burst", Label: "Signal burst effect", Kind: "visual"},
			},
		},
		{
			ID:                "side-step-spark",
			Title:             "Side Step Spark",
			Category:          "agility",
			Mechanic:          "side-steps",
			Description:       "Shift left and right across the capture lane to charge the spark meter.",
			TargetReps:        4,
			MinimumScore:      0.70,
			MinimumVisibility: 0.60,
			Cues: []string{
				"Move your hips clearly left and right from center.",
				"Keep the full body visible while changing sides.",
			},
			Unlocks: []Unlock{
				{ID: "lane-sparks", Label: "Lane spark overlay", Kind: "visual"},
			},
		},
	}
}

File: web/src/modules/catalog.ts
export type ModuleID = "motion-trail" | "wireframe-avatar" | "precision-export";

export type ModuleDefinition = {
  id: ModuleID;
  title: string;
  unlockIDs: string[];
  unlockChallengeID: string;
  description: string;
  lockedHint: string;
  status: "implemented";
};

export const moduleDefinitions: ModuleDefinition[] = [
  {
    id: "motion-trail",
    title: "Motion Trail",
    unlockIDs: ["trail-overlay", "lane-sparks"],
    unlockChallengeID: "calibrate-frame",
    description: "Draws a live center-of-body trace over the capture stage and module preview.",
    lockedHint: "Complete Frame Calibration to prove the tracker can see your full body.",
    status: "implemented"
  },
  {
    id: "wireframe-avatar",
    title: "Wireframe Avatar",
    unlockIDs: ["avatar-wireframe"],
    unlockChallengeID: "mirror-squat",
    description: "Mirrors pose quality, shoulder tilt, center line, and hand height as a simplified playable rig.",
    lockedHint: "Complete Mirror Squat to unlock the live avatar rig.",
    status: "implemented"
  },
  {
    id: "precision-export",
    title: "Precision Export",
    unlockIDs: ["precision-export"],
    unlockChallengeID: "balance-beacon",
    description: "Previews and downloads a session JSON summary with challenge, unlock, and pose-quality data.",
    lockedHint: "Complete Balance Beacon to unlock the export preview.",
    status: "implemented"
  }
];

export function isModuleUnlocked(module: ModuleDefinition, unlockedIDs: Set<string>): boolean {
  return module.unlockIDs.some((unlockID) => unlockedIDs.has(unlockID));
}
